### DISCUSSION TITLE
How to use Functions (documentation example doesn't work)

### DISCUSSION THREAD

**Discussion by kriskrej** [betalgo/openai#466] [General]
This line in the documentation on the main page:
"Functions = new List<FunctionDefinition> { fn1, fn2, fn3, fn4 },"
is not working anymore.

It seems like it should have been replaced with something like:
ToolChoice =  ToolChoice.FunctionChoice(...

but I wasn't able to get it to work or to find a working example on the web.
Help would be appreciated

**Comment by Drosos-kalyvas** [comment]
Same issue here, had to rollback to 7.3.1 to make the example work. Any help will be much appreciated

**Comment by kayhantolga** [comment]
There are some examples available in the playground. Are there any issues with them?

**Comment by kayhantolga** [comment]
I think I or OpenAI broke function calling. I will be focusing on fixing that this week. Additionally, I plan to provide additional documentation. Please leave a comment here if the playground sample doesn't work for you or if you have any feedback about function calling.

**Reply by Drosos-kalyvas** [reply]
Hello, using package version 7.3.1 function calling still works fine for now but only with GPT 3.5 Turbo and not with GPT 4. 
Also I am using .NET 5 so if Utilities will be required for function calling it is not going to work for me either way. But yeah the examples posted here
https://github.com/betalgo/openai/wiki/Function-Calling
don't work for the current version
Also, additional documentation will be amazing, thanks!

**Reply by Nicelittlewood** [reply]
switch to Tool is easier than you expect.

you have to set Tools like

Tools = new List<ToolDefinition> {td, td1, td2, td3 ... }

to Define a Tool:

ToolDefinition td = new ToolDefinition();
td.Type = StaticValues.CompletionStatics.ToolType.Function; //This define that the tool is a function 

td.Function = new FunctionDefinitionBuilder([Name], [Description]).AddParameter(....).... //define the function as like the old example

done.

**Reply by Drosos-kalyvas** [reply]
That does look pretty easy! I'll try this out today, thanks!

**Reply by Drosos-kalyvas** [reply]
@Nicelittlewood 
 Your example helped me very much, thanks a lot! Can you possible give me an example of the syntax to use                
 .AddParameter("array", PropertyDefinition.DefineArray()) and how to pass parameters inside? 
 For example I have this function to retun a flight request object:
```
td1.Function = new FunctionDefinitionBuilder("deserialize_response", "Deserialize response if it is a new flight request")
    .AddParameter("DepartureDate", PropertyDefinition.DefineString("The departure date. Use dd/mm/yyyy format. Example: 20/06/2024"))
    .AddParameter("From_AirportCode", PropertyDefinition.DefineString("The departure airport's IATA code. Must be code and not full name"))
    .AddParameter("To_AirportCode", PropertyDefinition.DefineString("The destination airport's IATA code. Must be code and not full name"))
    .AddParameter("DepartureTime", PropertyDefinition.DefineString("The time of departure flight"))
    .AddParameter("ReturnTime", PropertyDefinition.DefineString("The time of return flight"))
    .Validate()
    .Build();
```
How can I make it retun an array of this object?

**Reply by Nicelittlewood** [reply]
it should be like this:

```
PropertyDefinition pd = new PropertyDefinition();
pd.Description = "a fly request";
pd.Type = "object";

pd.Properties.Add("DepartureDate", PropertyDefinition.DefineString("The departure date. Use dd/mm/yyyy format. Example: 20/06/2024"));
pd.Properties.Add("DepartureTime", PropertyDefinition.DefineString("The time of departure flight"));
pd.Properties.Add("From_AirportCode", PropertyDefinition.DefineString("The departure airport's IATA code. Must be code and not full name"));
pd.Properties.Add("To_AirportCode", PropertyDefinition.DefineString("The destination airport's IATA code. Must be code and not full name"));
pd.Properties.Add("ReturnDate", PropertyDefinition.DefineString("The return date. Use dd/mm/yyyy format. Example: 20/06/2024"));
pd.Properties.Add("ReturnTime", PropertyDefinition.DefineString("The time of return flight"));


FunctionDefinition fd = new FunctionDefinitionBuilder("fly_requests","an array of fly request with detail").AddParameter("requests", PropertyDefinition.DefineArray(pd)).Validate().Build();
```

**Reply by Drosos-kalyvas** [reply]
Amazing!
Thanks so much

**Comment by kayhantolga** [comment]
I have solved a couple of bugs. I plan to merge them tomorrow and release a new version. 

As @Nicelittlewood mentioned, updating functions to tools is quite easy. Just imagine a tool as a wrapper of a function.
I believe that updating the current documentation will provide enough knowledge to start using the SDK. I highly encourage trying out the playground project, as I think it will be very helpful.

**Comment by kriskrej** [comment]
Thank you very much for your help! It works now.
